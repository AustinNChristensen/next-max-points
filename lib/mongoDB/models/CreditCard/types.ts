import { ObjectId } from 'mongodb';
import { TCategoryName } from '../Category/types';

export interface IBonusPointCategory {
    // TODO - deploy to vercel, add env vars to vercel
    categoryName: TCategoryName;
    multiplier: number;
}

export type TCardIssuer = 'Amex' | 'Chase' | 'Capital One';

export type TCardType = 'Personal' | 'Business';

export interface ICreditCardSchema {
    slug: string;
    displayName: string;
    issuer: TCardIssuer;
    type: TCardType;
    // default to 0
    annualFee?: number;
    // default to 1
    pointMultiplierValue?: number;
    bonusCategories: IBonusPointCategory[];
    // default to 0
    signUpBonus?: number;
};

export interface ICreditCardModel extends ICreditCardSchema {
    _id: ObjectId;
}

export interface IPostNewCreditCard extends Omit<ICreditCardSchema, 'slug'> {}