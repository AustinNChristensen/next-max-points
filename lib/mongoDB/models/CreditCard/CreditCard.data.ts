import { ICreditCardSchema, TCardIssuer, TCardType } from './types';

export const allCreditCardIssuers: TCardIssuer[] = ['Amex', 'Chase', 'Capital One'];

export const allCreditCardTypes: TCardType[] = ['Personal', 'Business'];

export const allCreditCardsTestData: ICreditCardSchema[] = [
    {
        slug: 'Chase-Sapphire-Reserve',
        displayName: 'Chase Sapphire Reserve',
        issuer: 'Chase',
        type: 'Personal',
        annualFee: 550,
        signUpBonus: 50000,
        pointMultiplierValue: 1.5,
        bonusCategories: [
            {
                categoryName: 'Dining',
                multiplier: 3
            },
            {
                categoryName: 'Travel',
                multiplier: 3
            },
            {
                categoryName: 'Everything Else',
                multiplier: 1
            }
        ]
    },
    {
        slug: 'Chase-Sapphire-Preferred',
        displayName: 'Chase Sapphire Preferred',
        issuer: 'Chase',
        type: 'Personal',
        annualFee: 95,
        signUpBonus: 80000,
        pointMultiplierValue: 1.25,
        bonusCategories: [
            {
                categoryName: 'Dining',
                multiplier: 2
            },
            {
                categoryName: 'Travel',
                multiplier: 2
            },
            {
                categoryName: 'Everything Else',
                multiplier: 1
            },
            {
                categoryName: 'Online Grocery',
                multiplier: 3
            },
            {
                categoryName: 'Streaming',
                multiplier: 3
            }
        ]
    },
    {
        slug: 'Chase-Freedom-Flex',
        displayName: 'Chase Freedom Flex',
        issuer: 'Chase',
        type: 'Personal',
        annualFee: 0,
        signUpBonus: 20000,
        pointMultiplierValue: 1,
        bonusCategories: [
            {
                categoryName: 'Dining',
                multiplier: 3
            },
            {
                categoryName: 'Drug Stores',
                multiplier: 3
            },
            {
                categoryName: 'Travel',
                multiplier: 5
            },
            {
                categoryName: 'Everything Else',
                multiplier: 1
            }
        ]
    },
    {
        slug: 'Chase-Freedom-Unlimited',
        displayName: 'Chase Freedom Unlimited',
        issuer: 'Chase',
        type: 'Personal',
        annualFee: 0,
        signUpBonus: 20000,
        pointMultiplierValue: 1,
        bonusCategories: [
            {
                categoryName: 'Gas',
                multiplier: 5
            },
            {
                categoryName: 'Travel',
                multiplier: 5
            },
            {
                categoryName: 'Everything Else',
                multiplier: 1.5
            },
            {
                categoryName: 'Dining',
                multiplier: 3
            },
            {
                categoryName: 'Drug Stores',
                multiplier: 3
            }
        ]
    },
    {
        slug: 'Amex-Platinum',
        displayName: 'Amex Platinum',
        issuer: 'Amex',
        type: 'Personal',
        annualFee: 695,
        signUpBonus: 150000,
        pointMultiplierValue: 1,
        bonusCategories: [
            {
                categoryName: 'Flights',
                multiplier: 5
            },
            {
                categoryName: 'Hotels',
                multiplier: 5
            },
            {
                categoryName: 'Everything Else',
                multiplier: 1
            }
        ]
    },
    {
        slug: 'Amex-Gold',
        displayName: 'Amex Gold',
        issuer: 'Amex',
        type: 'Personal',
        annualFee: 250,
        signUpBonus: 60000,
        pointMultiplierValue: 1,
        bonusCategories: [
            {
                categoryName: 'Dining',
                multiplier: 4
            },
            {
                categoryName: 'Flights',
                multiplier: 3
            },
            {
                categoryName: 'Everything Else',
                multiplier: 1
            },
            {
                categoryName: 'Groceries',
                multiplier: 4
            }
        ]
    },
    {
        slug: 'Amex-Green',
        displayName: 'Amex Green',
        issuer: 'Amex',
        type: 'Personal',
        annualFee: 150,
        signUpBonus: 40000,
        pointMultiplierValue: 1,
        bonusCategories: [
            {
                categoryName: 'Dining',
                multiplier: 3
            },
            {
                categoryName: 'Travel',
                multiplier: 3
            },
            {
                categoryName: 'Transit',
                multiplier: 3
            }
        ]
    },
    {
        slug: 'Amex-Blue-Cash EveryDay',
        displayName: 'Amex Blue Cash EveryDay',
        issuer: 'Amex',
        type: 'Personal',
        annualFee: 0,
        signUpBonus: 20000,
        pointMultiplierValue: 1,
        bonusCategories: [
            {
                categoryName: 'Groceries',
                multiplier: 3
            },
            {
                categoryName: 'Gas',
                multiplier: 2
            },
            {
                categoryName: 'Everything Else',
                multiplier: 1
            }
        ]
    },
    {
        slug: 'Amex-EveryDay',
        displayName: 'Amex EveryDay',
        issuer: 'Amex',
        type: 'Personal',
        annualFee: 0,
        signUpBonus: 10000,
        pointMultiplierValue: 1,
        bonusCategories: [
            {
                categoryName: 'Groceries',
                multiplier: 2
            },
            {
                categoryName: 'Everything Else',
                multiplier: 1
            }
        ]
    },
    {
        slug: 'Capital-One-Venture',
        displayName: 'Capital One Venture',
        issuer: 'Capital One',
        type: 'Personal',
        annualFee: 95,
        signUpBonus: 80000,
        pointMultiplierValue: 1.25,
        bonusCategories: [
            {
                categoryName: 'Everything Else',
                multiplier: 2
            },
            {
                categoryName: 'Hotels',
                multiplier: 5
            },
            {
                categoryName: 'Rental Cars',
                multiplier: 5
            }
        ]
    },
    {
        slug: 'Captial-One-Venture X',
        displayName: 'Captial One Venture X',
        issuer: 'Capital One',
        type: 'Personal',
        annualFee: 395,
        signUpBonus: 75000,
        pointMultiplierValue: 1,
        bonusCategories: [
            {
                categoryName: 'Travel',
                multiplier: 5
            },
            {
                categoryName: 'Everything Else',
                multiplier: 2
            },
            {
                categoryName: 'Rental Cars',
                multiplier: 10
            },
            {
                categoryName: 'Hotels',
                multiplier: 10
            }
        ]
    },
    {
        slug: 'Amex-Business-Gold',
        displayName: 'Amex Business Gold',
        issuer: 'Amex',
        type: 'Business',
        annualFee: 295,
        signUpBonus: 70000,
        pointMultiplierValue: 1,
        bonusCategories: [
            {
                categoryName: 'Advertising',
                multiplier: 2
            },
            {
                categoryName: 'Computer Hardware',
                multiplier: 2
            },
            {
                categoryName: 'Everything Else',
                multiplier: 1
            }
        ]
    },
    {
        slug: 'Amex-Business-Platinum',
        displayName: 'Amex Business Platinum',
        issuer: 'Amex',
        type: 'Business',
        annualFee: 695,
        signUpBonus: 120000,
        pointMultiplierValue: 1,
        bonusCategories: [
            {
                categoryName: 'Flights',
                multiplier: 5
            },
            {
                categoryName: 'Everything Else',
                multiplier: 2
            },
            {
                categoryName: 'Hotels',
                multiplier: 5
            }
        ]
    },
    {
        slug: 'Chase-Ink-Business Preferred',
        displayName: 'Chase Ink Business Preferred',
        issuer: 'Chase',
        type: 'Business',
        annualFee: 95,
        signUpBonus: 100000,
        pointMultiplierValue: 1.25,
        bonusCategories: [
            {
                categoryName: 'Shipping',
                multiplier: 3
            },
            {
                categoryName: 'Advertising',
                multiplier: 3
            },
            {
                categoryName: 'Internet',
                multiplier: 3
            },
            {
                categoryName: 'Everything Else',
                multiplier: 1
            }
        ]
    },
    {
        slug: 'Chase-Ink-Business Cash',
        displayName: 'Chase Ink Business Cash',
        issuer: 'Chase',
        type: 'Business',
        annualFee: 0,
        signUpBonus: 75000,
        pointMultiplierValue: 1,
        bonusCategories: [
            {
                categoryName: 'Office Supply',
                multiplier: 5
            },
            {
                categoryName: 'Everything Else',
                multiplier: 1
            }
        ]
    },
    {
        slug: 'Chase-Ink-Business Unlimited',
        displayName: 'Chase Ink Business Unlimited',
        issuer: 'Chase',
        type: 'Business',
        annualFee: 0,
        signUpBonus: 75000,
        pointMultiplierValue: 1,
        bonusCategories: [
            {
                categoryName: 'Everything Else',
                multiplier: 1.5
            }
        ]
    }
];