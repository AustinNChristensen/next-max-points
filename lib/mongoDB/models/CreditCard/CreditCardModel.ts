import { models, model, Schema } from 'mongoose';
import { allBonusPointCategories } from '../Category/Category.data';
import { allCreditCardIssuers, allCreditCardTypes } from './CreditCard.data';
import { ICreditCardSchema, IBonusPointCategory } from './types';

const BonusCategorySchema = new Schema<IBonusPointCategory>({
    categoryName: {
        type: String,
        enum: [...allBonusPointCategories]
    },
    multiplier: {
        type: Number,
        default: 1
    }
});

export const CreditCardSchema = new Schema<ICreditCardSchema>({
    slug: {
        type: String
    },
    displayName: {
        type: String
    },
    issuer: {
        type: String,
        enum: [...allCreditCardIssuers]
    },
    type: {
        type: String,
        enum: [...allCreditCardTypes]
    },
    annualFee: {
        type: Number,
        default: 0
    },
    pointMultiplierValue: {
        type: Number,
        default: 1
    },
    bonusCategories: [BonusCategorySchema],
    signUpBonus: {
        type: Number,
        default: 0
    }
});

export const CreditCardModel = models.CreditCard || model('CreditCard', CreditCardSchema);