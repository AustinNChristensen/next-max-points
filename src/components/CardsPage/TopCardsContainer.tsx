import { TCategoryName } from '../../../lib/mongoDB/models/Category/types';
import { ICreditCardModel } from '../../../lib/mongoDB/models/CreditCard/types';

interface Props {
    topCardsByCategory: {
        categoryName: TCategoryName;
        cards: ICreditCardModel
    }[];
}

export const TopCardsContainer = ({ topCardsByCategory }: Props) => {
    return (
        <div className='px-4 sm:px-6 lg:px-8'>
            <div className='sm:flex sm:items-center'>
                <div className='sm:flex-auto'>
                    <h1 className='text-xl font-semibold text-gray-900'>Top Cards</h1>
                    <p className='mt-2 text-sm text-gray-700'>
                        A list of top Credit Cards by Category
                    </p>
                </div>
            </div>
            <div className='mt-8 flex flex-col'>
                <div className='-my-2 -mx-4 overflow-x-auto sm:-mx-6 lg:-mx-8'>
                    <div className='inline-block min-w-full py-2 align-middle md:px-6 lg:px-8'>
                        <div className='overflow-hidden md:rounded-lg'>
                            <ul
                                role='list'
                                className='grid grid-cols-1 gap-6 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4'
                            >
                                {topCardsByCategory.map((category) => {
                                    const card = category.cards;
                                    const matchingCategory = card.bonusCategories.find((cat) => cat.categoryName === category.categoryName);
                                    const categoryEarnRate = (matchingCategory?.multiplier || 1) * (card.pointMultiplierValue || 1);
                                    return (
                                        <li
                                            key={card._id.toString()}
                                            className='col-span-1 flex flex-col text-center bg-white rounded-lg shadow divide-y divide-gray-200'
                                        >
                                            <div className='flex-1 flex flex-col p-8'>
                                                <h3 className='mt-6 text-gray-900 text-md font-bold'>{category.categoryName}</h3>
                                                <h3 className='mt-6 text-gray-900 text-sm font-medium'>{card.displayName}</h3>
                                                <dl className='mt-1 flex-grow flex flex-col justify-between'>
                                                    <dd className='text-gray-500 text-sm'>{card.issuer}</dd>
                                                    <dd className='text-gray-500 text-sm'>{card.type}</dd>
                                                    <dd className='mt-3'>
                                                        <span className='px-2 py-1 text-green-800 text-xs font-medium bg-green-100 rounded-full'>
                                                            {`${categoryEarnRate}%`}
                                                        </span>
                                                    </dd>
                                                </dl>
                                            </div>
                                        </li>
                                    );
                                })}
                            </ul>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    );
};