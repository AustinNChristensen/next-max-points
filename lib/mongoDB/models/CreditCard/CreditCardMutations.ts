import { kebabCase } from 'lodash';
import { ObjectId } from 'mongodb';
import { PartialDeep } from 'type-fest';
import MongoClient from '../../client';
import { TCategoryName } from '../Category/types';
import { allCreditCardsTestData } from './CreditCard.data';
import { CreditCardModel } from './CreditCardModel';
import { IBonusPointCategory, ICreditCardModel, ICreditCardSchema, IPostNewCreditCard } from './types';

export const seedCreditCardCollection = async () => {
    MongoClient;

    await CreditCardModel.deleteMany();

    return await CreditCardModel.insertMany(allCreditCardsTestData);
};

export const createNewCreditCard = async (cardData: IPostNewCreditCard): Promise<ICreditCardModel> => {
    MongoClient;

    const newCard: ICreditCardSchema = {
        ...cardData,
        slug: kebabCase(cardData.displayName)
    };

    return await CreditCardModel.create(newCard);
};

export const getAllCreditCards = async (): Promise<ICreditCardModel[]> => {
    MongoClient;
    return await CreditCardModel.find().exec();
};

export const getCreditCardByID = async (cardID: string | ObjectId): Promise<ICreditCardModel> => {
    MongoClient;
    return await CreditCardModel.findById(cardID.toString()).exec();
};

export const updateCreditCardByID = async (cardID: string | ObjectId, data: PartialDeep<ICreditCardModel>): Promise<ICreditCardModel> => {
    MongoClient;
    return await CreditCardModel.findByIdAndUpdate(cardID.toString(), data, { returnDocument: 'after' }).exec();
};

export const deleteCreditCardByID = async (cardID: string | ObjectId): Promise<ICreditCardModel> => {
    MongoClient;
    return await CreditCardModel.findByIdAndDelete(cardID.toString()).exec();
};

export const getBonusCategoriesByCreditCardID = async (cardID: string | ObjectId): Promise<IBonusPointCategory[]> => {
    MongoClient;
    return (await CreditCardModel.findById(cardID.toString()).select('bonusCategories').exec()).bonusCategories;
};

export const getTopCreditCardsByCategory = async (categoryID: TCategoryName): Promise<ICreditCardModel[]> => {
    MongoClient;
    return await CreditCardModel.find({ 'bonusCategories.categoryName': categoryID }).sort(`bonusCategories.multiplier`).limit(25).exec();
};