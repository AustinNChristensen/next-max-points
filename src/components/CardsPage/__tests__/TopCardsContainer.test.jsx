/**
 * @jest-environment jsdom
 */
import { render, screen } from '@testing-library/react';
import { TopCardsContainer } from '../TopCardsContainer';
import { allCreditCardsTestData } from '../../../../lib/mongoDB/models/CreditCard/CreditCard.data';

describe('Top Cards Container', () => {
    let topCardsByCategory;

    beforeEach(() => {
        topCardsByCategory = [
            {
                categoryName: 'Everything Else',
                cards: {
                    _id: 'testID',
                    ...allCreditCardsTestData[0]
                }
            }
        ];
    });

    const setupRTL = () => render(<TopCardsContainer topCardsByCategory={topCardsByCategory} />);

    test('matches snapshot', () => {
        const { container } = setupRTL();

        expect(container).toMatchSnapshot();
    });

    test('renders page text', () => {
        setupRTL();

        expect(screen.getByRole('heading', { name: 'Top Cards' })).toBeInTheDocument();
        expect(screen.getByText('A list of top Credit Cards by Category')).toBeInTheDocument();

        const testCard = allCreditCardsTestData[0];

        expect(screen.getByText(testCard.displayName)).toBeInTheDocument();
        expect(screen.getByText(testCard.issuer)).toBeInTheDocument();
        expect(screen.getByText(testCard.type)).toBeInTheDocument();
        expect(screen.getByText('1.5%')).toBeInTheDocument();
    });
});