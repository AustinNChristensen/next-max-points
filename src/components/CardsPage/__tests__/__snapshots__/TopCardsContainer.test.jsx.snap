// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Top Cards Container matches snapshot 1`] = `
<div>
  <div
    class="px-4 sm:px-6 lg:px-8"
  >
    <div
      class="sm:flex sm:items-center"
    >
      <div
        class="sm:flex-auto"
      >
        <h1
          class="text-xl font-semibold text-gray-900"
        >
          Top Cards
        </h1>
        <p
          class="mt-2 text-sm text-gray-700"
        >
          A list of top Credit Cards by Category
        </p>
      </div>
    </div>
    <div
      class="mt-8 flex flex-col"
    >
      <div
        class="-my-2 -mx-4 overflow-x-auto sm:-mx-6 lg:-mx-8"
      >
        <div
          class="inline-block min-w-full py-2 align-middle md:px-6 lg:px-8"
        >
          <div
            class="overflow-hidden md:rounded-lg"
          >
            <ul
              class="grid grid-cols-1 gap-6 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4"
              role="list"
            >
              <li
                class="col-span-1 flex flex-col text-center bg-white rounded-lg shadow divide-y divide-gray-200"
              >
                <div
                  class="flex-1 flex flex-col p-8"
                >
                  <h3
                    class="mt-6 text-gray-900 text-md font-bold"
                  >
                    Everything Else
                  </h3>
                  <h3
                    class="mt-6 text-gray-900 text-sm font-medium"
                  >
                    Chase Sapphire Reserve
                  </h3>
                  <dl
                    class="mt-1 flex-grow flex flex-col justify-between"
                  >
                    <dd
                      class="text-gray-500 text-sm"
                    >
                      Chase
                    </dd>
                    <dd
                      class="text-gray-500 text-sm"
                    >
                      Personal
                    </dd>
                    <dd
                      class="mt-3"
                    >
                      <span
                        class="px-2 py-1 text-green-800 text-xs font-medium bg-green-100 rounded-full"
                      >
                        1.5%
                      </span>
                    </dd>
                  </dl>
                </div>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
